### PR TITLE
Restrict calling langlib functions for objects

### LINKED ISSUE #23944: Disallow the use of method call expressions for calling lang lib functions for objects
As James pointed out in https://github.com/ballerina-platform/ballerina-spec/issues/543, calling lang lib functions using method call expressions should not be allowed for objects. Currently we allow usages such as the following:
```ballerina
type Person object {
    string name = "";
    int age = 0;
};

public function main() {
    Person p = new;
    string s = p.toString(); // This should not be allowed
}
```

### LINKED ISSUE #29958: Included methods not getting populated as referenced functions
**Description:**
At BIRPackageSymbolEnter, the included methods are not getting populated under the symbol.

**Steps to reproduce:**
```ballerina
public type Frame object {
   public function toString() returns string;
};

public type Frame2 object {
    *Frame;
};
```

Above should be loaded through BIR.

```ballerina
public class FrameImpl {
   *Frame2;  // this should throw an error as toString() is not implemented
}
```

**Affected Versions:**
SwanLake 4 

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by pcnfernando**
Fixes #23944
Fixes #29958

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by pcnfernando** [comment]
BIRSpec tests are failing because of https://github.com/ballerina-platform/ballerina-lang/issues/29958.

**Comment by IMS94** [comment]
@pcnfernando We had a similar requirement at language server and did https://github.com/ballerina-platform/ballerina-lang/pull/29873. How does that stack up with this PR?

**Comment by pcnfernando** [comment]
> @pcnfernando We had a similar requirement at language server and did #29873. How does that stack up with this PR?

With the said fix, you have filtered out the suggestions. That change is anyway needed. With the change I'm introducing the usage is restricted from the compiler.

**Comment by IMS94** [comment]
@pcnfernando Thanks! We did that at the `BallerinaObjectTypeSymbol` level like this: https://github.com/ballerina-platform/ballerina-lang/pull/29873/files#diff-5855e99980da5c8dbbaaf324f3d368f96e29a94b97233baefe9f083f8cf32b2e. Hope that's ok.

**Comment by MaryamZi** [comment]
@pcnfernando there seems to be a conflict :(

File: tests/jballerina-unit-test/src/test/resources/test-src/object/object_langlib_function_invocation_negative.bal [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import ballerina/lang.'object as lang_obj;
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/object/object_langlib_function_invocation_negative.bal:1]: Missing the licence header.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java [hunk 2]
```diff
@@ -255,10 +255,42 @@ private BPackageSymbol definePackage(DataInputStream dataInStream, int pkgCpInde
         // Define service declarations
         defineSymbols(dataInStream, rethrow(this::defineServiceDeclarations));
 
+        populateReferencedFunctions();
+
         this.typeReader = null;
         return this.env.pkgSymbol;
     }
 
+    private void populateReferencedFunctions() {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java:264]: Have we added tests for this change? Shall we add both positive and negative bala tests?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java [hunk 3]
```diff
@@ -1453,7 +1485,31 @@ public BType readType(int cpI) throws IOException {
...
+                        //populate intersection type object functions
+                        if (isImmutable(objectSymbol.flags) && Symbols.isFlagOn(flags, Flags.ANONYMOUS)) {
+                            String attachedFuncName = getStringCPEntryValue(inputStream);
+                            var attachedFuncFlags = inputStream.readLong();
+                            if (Symbols.isFlagOn(attachedFuncFlags, Flags.INTERFACE) &&
+                                    Symbols.isFlagOn(attachedFuncFlags, Flags.ATTACHED)) {
+                                BInvokableType attachedFuncType = (BInvokableType) readTypeFromCp();
+                                Name funcName = names.fromString(Symbols.getAttachedFuncSymbolName(
+                                        objectSymbol.name.value, attachedFuncName));
+                                BInvokableSymbol attachedFuncSymbol =
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java:1497]: When an implementation is not found for a function, the function name in the error seems to have additional info when it is via the BIR.

```ballerina
import ballerina/lang.runtime;

public type StackFrame readonly & object {
   public function toString() returns string;
};

readonly class Foo {
    *StackFrame; // no implementation found for the method 'toString' of class 'Foo'
}

readonly class Bar {
    *runtime:StackFrame; // no implementation found for the method '(ballerina/lang.runtime:0.0.1:$anonType$_0 & readonly).toString' of class 'Bar'
}
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java [hunk 4]
```diff
@@ -1453,7 +1485,31 @@ public BType readType(int cpI) throws IOException {
                     }
                     int funcCount = inputStream.readInt();
                     for (int i = 0; i < funcCount; i++) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java:1487]: Shall we extract this out to a method?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java [hunk 5]
```diff
@@ -255,10 +255,42 @@ private BPackageSymbol definePackage(DataInputStream dataInStream, int pkgCpInde
...
 
+        populateReferencedFunctions();
+
         this.typeReader = null;
         return this.env.pkgSymbol;
     }
 
+    private void populateReferencedFunctions() {
+        for (BStructureTypeSymbol structureTypeSymbol : this.structureTypes) {
+            if (structureTypeSymbol instanceof BObjectTypeSymbol) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java:266]: Can't/shouldn't we check `structureTypeSymbol.type.tag` instead? Or at least symbol kind.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java [hunk 6]
```diff
@@ -255,10 +255,42 @@ private BPackageSymbol definePackage(DataInputStream dataInStream, int pkgCpInde
...
+                        Name funcName = names.fromString(
+                                Symbols.getAttachedFuncSymbolName(structureTypeSymbol.name.value, referencedFuncName));
+                        Scope.ScopeEntry matchingObjFuncSym = this.env.pkgSymbol.scope.lookup(funcName);
+                        if (matchingObjFuncSym != null) {
+                            continue;
+                        }
+                        if (Symbols.isPrivate(function.symbol)) {
+                            // we should not copy private functions.
+                            continue;
+                        }
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java:285]: Shouldn't this be the first thing we do within the for loop?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java (hunk 2 above)
Comment (pcnfernando) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java:264]: This is covered under object_langlib_function_invocation_negative.bal

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java:264]: That doesn't test the BIR symbol enter, right? We need to add a bala test?

Comment (pcnfernando) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java:264]: Added tests in 1b6b21ff9a8ad716b114fd76ad04f19dc70a57e8

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java (hunk 4 above)
Comment (pcnfernando) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java:1487]: Refactored in 1b6b21ff9a8ad716b114fd76ad04f19dc70a57e8

File: tests/jballerina-unit-test/src/test/resources/test-src/object/object_langlib_function_invocation_negative.bal (hunk 1 above)
Comment (pcnfernando) [reply tests/jballerina-unit-test/src/test/resources/test-src/object/object_langlib_function_invocation_negative.bal:17]: Added in 1b6b21ff9a8ad716b114fd76ad04f19dc70a57e8

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java (hunk 3 above)
Comment (pcnfernando) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java]: Addressed in 1b6b21ff9a8ad716b114fd76ad04f19dc70a57e8

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java (hunk 5 above)
Comment (pcnfernando) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java]: Updated in 1b6b21ff9a8ad716b114fd76ad04f19dc70a57e8

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java (hunk 6 above)
Comment (pcnfernando) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java]: Updated in 1b6b21ff9a8ad716b114fd76ad04f19dc70a57e8

**Review by gimantha — APPROVED** [review]

File: tests/jballerina-unit-test/src/test/resources/test-src/object/object_langlib_function_invocation_negative.bal [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+    public function immediateStop() returns error? {
+        return ();
+    }
+}
+
+public readonly class StackFrameImpl {
+    *runtime:StackFrame;
+}
+
+public readonly class StackFrameImpl2 {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/object/object_langlib_function_invocation_negative.bal:92]: Here onward, it's a positive test right? Shall we move this to a separate file?

And IMO it will be better if we can introduce a separate project and use constructs from that rather than (or in addition to) using langlib modules like we do with other bala tests, so that the user-defined project path is also tested.

We can do this in a separate PR.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java (hunk 2 above)
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java:264]: Please check the following sample, I do not get an error for not having an `xyz` method in `MyBar`.

foo module
```ballerina
public type FooObj object {
    public function xyz();
};

public type BarObj object {
    *FooObj;
    public function abc();
};
```

Module that imports `foo`
```ballerina
class MyBar {
    *foo:BarObj;

    public function abc() {} // should result in an error for missing `xyz`
}
```

Comment (pcnfernando) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java:265]: Added a bala test in 2a8045f1513eb0a447d358b96b77ad6ab656a715

File: tests/jballerina-unit-test/src/test/resources/test-src/object/object_langlib_function_invocation_negative.bal (hunk 7 above)
Comment (pcnfernando) [reply tests/jballerina-unit-test/src/test/resources/test-src/object/object_langlib_function_invocation_negative.bal]: Refactored to separate files and improved and enabled bala test in 2a8045f1513eb0a447d358b96b77ad6ab656a715

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java [hunk 8]
```diff
@@ -1453,7 +1485,12 @@ public BType readType(int cpI) throws IOException {
                     }
                     int funcCount = inputStream.readInt();
                     for (int i = 0; i < funcCount; i++) {
-                        ignoreAttachedFunc();
+                        //populate intersection type object functions
+                        if (isImmutable(objectSymbol.flags) && Symbols.isFlagOn(flags, Flags.ANONYMOUS)) {
+                            populateIntersectionTypeReferencedFunctions(inputStream, objectSymbol);
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java:1490]: Shall we create an issue for these methods not being available as top level functions also? We need to verify if they should be added.

**Review by MaryamZi — APPROVED** [review]

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java (hunk 8 above)
Comment (pcnfernando) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java:1490]: https://github.com/ballerina-platform/ballerina-lang/issues/30051 tracks this
